include/nbl/asset/utils/CSPIRVIntrospector.h

```diff
@@ -13,172 +17,657 @@
+						{
+							arraySizes[i].specID = sizes[i]; // ID of spec constant if size is spec constant
+							arraySizes[i].isSpecConstant = true;
+						}
+					}
+
+					return range;
+				}
+				inline core::based_span<char> addString(const std::string_view str) // TODO: move to cpp
+				{
+					const auto range = alloc<char>(str.size()+1);
+					const auto out = range(m_memPool.data());
+					memcpy(out.data(),str.data(),str.size());
+					out.back() = '\0';
+					return range;
+				}
+				SDescriptorVarInfo<true>* addResource(const spirv_cross::Compiler& comp, const spirv_cross::Resource& r, IDescriptor::E_TYPE restype);
+				inline core::based_offset<SType<true>> addType(const size_t memberCount) // TODO: move to cpp
+				{
+					const auto memberStorage = allocOffset(SType<true>::StoragePerMember*memberCount);
+					auto retval = alloc<SType<true>>(1);
+					// allocated everything before touching any data stores, to avoid pointer invalidation
+					auto pType = retval(m_memPool.data()).data();
+					pType->memberCount = memberCount;
+					pType->memberInfoStorage = {memberStorage};
+					return {retval.byte_offset()};
+				}
+				void shaderMemBlockIntrospection(const spirv_cross::Compiler& comp, SMemoryBlock<true>* root, const spirv_cross::Resource& r);
+				void finalize(IShader::E_SHADER_STAGE stage);
+
+				//! debug
+				static void printExtents(std::ostringstream& out, const std::span<const SArrayInfo> counts);
+				static void printType(std::ostringstream& out, const SType<false>* counts, const uint32_t depth=0);
+				
+				IShader::E_SHADER_STAGE m_shaderStage = IShader::E_SHADER_STAGE::ESS_UNKNOWN;
+
+				// Parameters it was created with
+				SParams m_params;
+				//! Sorted by `id`
+				core::vector<SSpecConstant<>> m_specConstants; // TODO: maybe unordered_set?
+				//! Sorted by `location`
+				core::vector<SInputInterface> m_input;
+				std::variant<
+					core::vector<SFragmentOutputInterface>, // when `params.shader->getStage()==ESS_FRAGMENT`
+					core::vector<SOutputInterface> // otherwise
+				> m_output;
+				//!
+				SPushConstantInfo<> m_pushConstants = {};
+				//! Each vector is sorted by `binding`
+				constexpr static inline uint8_t DESCRIPTOR_SET_COUNT = 4;
+				core::vector<SDescriptorVarInfo<>> m_descriptorSetBindings[DESCRIPTOR_SET_COUNT];
+				// The idea is that we construct with based_span (so we can add `.data()` when accessing)
+				// then just convert from `SMemoryBlock<true>` to `SMemoryBlock<false>`
+				// in-place when filling in the `descriptorSetBinding` vector which we pass to ctor
+				core::vector<char> m_memPool;
+		};
+		class CPipelineIntrospectionData final : public CIntrospectionData
 		{
-			std::string entryPoint;
-			core::smart_refctd_ptr<const ICPUShader> cpuShader;
+			public:
+				struct SDescriptorInfo final : CIntrospectionData::SDescriptorInfo
+				{
+					inline bool isArray() const {return stride;}
+					inline bool isRuntimeSized() const {return isArray() && count==0;}
```

Comment:
if there's no iff relationship https://discord.com/channels/427551838099996672/427951526967902218/1230618569931227146 then we need to track `isruntimeSized` separately from `count` as `count` would just become a lower bound on the size of the array then